File: CPP/test_deck.cpp
#include "deck.h"

#include "gtest/gtest.h"

TEST(Card, init)
{
    Card *c = new Card(4, Suit::Diamonds);
    EXPECT_EQ(c->get_rank(), 4);
    EXPECT_EQ(c->get_suit(), Suit::Diamonds);
}

TEST(Deck, init)
{
    Deck *d = new Deck();
    EXPECT_EQ(d->size(), 55);
}

TEST(Card, operators)
{
    Card c1(5, Suit::Clubs);
    Card c2(6, Suit::Diamonds);
    Card c3(5, Suit::Hearts);
    EXPECT_TRUE(c3 == c1);
    EXPECT_FALSE(c2 == c1);
    EXPECT_TRUE(c1 < c2);
    EXPECT_FALSE(c3 < c1);
    EXPECT_TRUE(c2 > c3);
    EXPECT_FALSE(c3 > c1);
}

TEST(Deck, size)
{
    Deck *d = new Deck();
    EXPECT_EQ(d->size(), 55);
}

TEST(Deck, sort)
{
    Deck *d = new Deck();
    d->sort();
    Card c = d->get_cards()[4];
    EXPECT_EQ(c.get_rank(), 1);
    c = d->get_cards()[19];
    EXPECT_EQ(c.get_rank(), 5);
    c = d->get_cards()[52];
    EXPECT_EQ(c.get_rank(), 13);
}

TEST(Deck, take)
{
    Deck *d = new Deck();
    EXPECT_EQ(d->take().get_rank(), 0);
    d->sort();
    EXPECT_EQ(d->take().get_rank(), 13);
}

TEST(Deck, put)
{
    Deck *d = new Deck();
    Card c1(5, Suit::Clubs);
    Card c2(6, Suit::Diamonds);
    Card c3(13, Suit::Hearts);
    d->put(c3);
    d->put(c2);
    d->put(c1);
    EXPECT_EQ(d->take().get_rank(), 5);
    EXPECT_EQ(d->take().get_rank(), 6);
    EXPECT_EQ(d->take().get_rank(), 13);
}

TEST(Deck, pick_by_random)
{
    Deck *d = new Deck();
    Card c1 = d->pick_by_random();
    EXPECT_TRUE(c1.get_rank() >= 0);
}

TEST(Deck, sort_by_value)
{
    Deck *d = new Deck();
    d->sort_by_value();
    Card c1 = d->get_cards()[3];
    EXPECT_EQ(c1.get_rank(), 1);
    EXPECT_EQ(c1.get_suit(), Suit::Hearts);
    c1 = d->get_cards()[4];
    EXPECT_EQ(c1.get_rank(), 1);
    EXPECT_EQ(c1.get_suit(), Suit::Diamonds);
    c1 = d->get_cards()[19];
    EXPECT_EQ(c1.get_rank(), 5);
    EXPECT_EQ(c1.get_suit(), Suit::Hearts);
}

TEST(Deck, sort_by_suit)
{
    Deck *d = new Deck();
    d->sort_by_suit();
    Card c1 = d->get_cards()[0];
    EXPECT_EQ(c1.get_rank(), 1);
    EXPECT_EQ(c1.get_suit(), Suit::Hearts);
    c1 = d->get_cards()[14];
    EXPECT_EQ(c1.get_rank(), 2);
    EXPECT_EQ(c1.get_suit(), Suit::Diamonds);
    c1 = d->get_cards()[29];
    EXPECT_EQ(c1.get_rank(), 4);
    EXPECT_EQ(c1.get_suit(), Suit::Spades);
}

TEST(Deck, remove_jokers)
{
    Deck *d = new Deck();
    d->remove_jokers();
    EXPECT_TRUE(d->take().get_rank() != 0);
    Deck *d2 = new Deck();
    d2->sort();
    d2->remove_jokers();
    Card c1 = d2->get_cards()[0];
    EXPECT_TRUE(c1.get_rank() != 0);
}

TEST(Deck, remove_duplicates)
{
    Deck *d = new Deck();
    vector<int> temp = d->remove_duplicates();
    EXPECT_TRUE(temp.size() == 0);
    Card c1(1, Suit::Diamonds);
    d->put(c1);
    temp = d->remove_duplicates();
    EXPECT_FALSE(temp.size() == 0);
}

TEST(Deck, insert)
{
    vector<Card> temp;
    Deck d;
    Deck *d2 = new Deck();
    d.insert(temp, d2->take());
    d.insert(temp, d2->take());
    d.insert(temp, d2->take());
    d.insert(temp, d2->take());
    d.insert(temp, d2->take());
    d.set_cards(temp);
    EXPECT_EQ(d.get_cards()[0].get_rank(), 0);
    EXPECT_EQ(d.get_cards()[3].get_rank(), 12);
}

TEST(Deck, deal)
{
    Deck *d = new Deck();
    Deck *d2 = new Deck();
    d->deal(4);
    EXPECT_EQ(d->size(), 3);
    d2->deal(5);
    EXPECT_EQ(d2->size(), 0);
}

File: Python/test_deck.py
from deck import Deck, Card


def test_card_init():
    # Test creating a card to see if the card is correctly created
    c = Card(5, 'Hearts')
    assert c.rank == 5
    assert c.suite == 'Hearts'

    print('Passed card init test')


def test_card_operators():
    # Test comparing card rank to see if the operators are functioning correctly
    c1 = Card(5, 'Hearts')
    c2 = Card(5, 'Spades')
    c3 = Card(4, 'Clubs')
    assert c1 == c2
    assert c3 < c1
    assert c1 != c3
    assert c2 > c3

    print('Passed card operators test')


def test_deck_init_len():
    # Test creating card deck to see if the right amount of cards are created
    # and if len function returns the right amount of cards
    d = Deck()
    assert len(d) == 55

    print('Passed deck init and len test')


def test_deck_sort():
    # Test sorting the card deck to see if the deck is sorted successfully
    d = Deck()
    d.sort()
    assert d.cards[1].rank == 0
    assert d.cards[4].rank == 1
    assert d.cards[53].rank == 13

    print('Passed deck sort test')


def test_deck_take():
    # Test taking cards from the top of the card deck to see if it is taking the correct card
    d = Deck()
    assert d.take().rank == 0
    d.sort()
    assert d.take().rank == 13
    assert len(d) == 53

    print('Passed deck take test')


def test_deck_put():
    # Test putting cards on top of the card deck to see if it is putting in the right cards
    d = Deck()
    c1 = Card(5, 'Hearts')
    c2 = Card(5, 'Spades')
    c3 = Card(4, 'Clubs')
    d.put(c1)
    d.put(c2)
    d.put(c3)
    assert d.take().rank == 4
    assert d.take().rank == 5
    assert d.take().rank == 5
    assert len(d) == 55

    print('Passed deck put test')


def test_deck_insert():
    # Test inserting cards in to a list
    test = []
    d = Deck()
    d.insert(test, d.take())
    assert len(test) == 1
    assert test[0].rank == 0
    d.sort()
    d.insert(test, d.take())
    assert len(test) == 2
    assert test[1].rank == 13

    print('Passed deck insert test')


def test_deck_remove_joker():
    # Test removing jokers to see if the jokers at the end of the decks are removed successfully
    d = Deck()
    assert len(d) == 55
    d.remove_jokers()
    assert len(d) == 52
    assert d.take().rank != 0

    print('Passed deck remove joker test')


def test_deck_remove_duplicates():
    # Test removing duplicates to see if any duplicates is found and removed successfully
    d = Deck()
    test = d.remove_duplicates()
    assert len(d) == 55
    assert len(test) == 0
    c = Card(5, 'Hearts')
    d.put(c)
    assert len(d) == 56
    test = d.remove_duplicates()
    assert len(d) == 55
    assert len(test) > 0

    print('Passed deck remove duplicates test')


def test_deck_pick_by_random():
    # Test picking a random card out of the deck and check if the card is picked out successfully
    d = Deck()
    assert len(d) == 55
    c1 = d.pick_by_random()
    assert len(d) == 54
    assert c1.rank >= 0
    c2 = d.pick_by_random()
    assert len(d) == 53
    assert c2.rank >= 0

    print('Passed deck pick by random test')


def test_deck_sort_by_value():
    # Test sorting the deck by value and checking if the deck is sorted successfully
    d = Deck()
    d.sort_by_value()
    c = d.cards[3]
    assert c.rank == 1
    assert c.suite == 'Hearts'
    c = d.cards[7]
    assert c.rank == 2
    assert c.suite == 'Hearts'
    c = d.cards[19]
    assert c.rank == 5
    assert c.suite == 'Hearts'

    print('Passed deck sort by value test')


def test_deck_sort_by_suit():
    # Test sorting the deck by suit and checking if the deck is sorted successfully
    d = Deck()
    d.sort_by_suit()
    c = d.cards[0]
    assert c.rank == 1
    assert c.suite == 'Hearts'
    c = d.cards[2]
    assert c.rank == 3
    assert c.suite == 'Hearts'
    c = d.cards[13]
    assert c.rank == 1
    assert c.suite == 'Diamonds'

    print('Passed deck sort by suit test')


def test_deck_deal():
    # Test dealing out cards to players to see if the cards are dealt out equally
    d1 = Deck()
    d1.deal(5)
    assert len(d1) == 0
    d2 = Deck()
    d2.deal(4)
    assert len(d2) > 0

    print('Passed deck deal test')


def run_all_card_test():
    # Run all test functions for Card
    print('---------------------\nTesting Card\n---------------------')
    test_card_init()
    test_card_operators()
    print('---------------------\nAll card test passed\n---------------------')


def run_all_deck_test():
    # Run all test functions for Deck
    print('---------------------\nTesting Deck\n---------------------')
    test_deck_init_len()
    test_deck_sort()
    test_deck_take()
    test_deck_put()
    test_deck_insert()
    test_deck_remove_joker()
    test_deck_remove_duplicates()
    test_deck_pick_by_random()
    test_deck_sort_by_value()
    test_deck_sort_by_suit()
    test_deck_deal()
    print('---------------------\nAll deck test passed\n---------------------')


def run_all_test():
    # Run all test functions
    run_all_card_test()
    run_all_deck_test()


if __name__ == '__main__':
    # Run all the test and if it passes, print out that it passes
    # if any test does not pass, it will not print out the test that fails and will stop the whole process
    run_all_test()
